src/coreclr/nativeaot/System.Private.CoreLib/src/System/GC.NativeAot.cs

```diff
@@ -274,6 +274,47 @@ public static bool TryStartNoGCRegion(long totalSize)
             return StartNoGCRegionWorker(totalSize, false, 0, false);
         }
 
+        internal static Action? s_NoGCCallback;
+
+        internal static void FinalizerAction(FinalizerActionKind finalizerActionKind)
+        {
+            Action? callback = s_NoGCCallback;
+            if ((finalizerActionKind & FinalizerActionKind.DetachNoGCCallback) != 0)
+            {
+                s_NoGCCallback = null;
+            }
+            if ((finalizerActionKind & FinalizerActionKind.ExecuteNoGCCallback) != 0)
+            {
+                callback?.Invoke();
+            }
+        }
+
+        public static unsafe void RegisterNoGCRegionCallback(long totalSize, Action callback)
+        {
+            ArgumentOutOfRangeException.ThrowIfNegativeOrZero(totalSize);
+            ArgumentNullException.ThrowIfNull(callback);
+
+            if (Interlocked.CompareExchange(ref s_NoGCCallback, callback, null) != null)
+            {
+                throw new InvalidOperationException(SR.InvalidOperationException_NoGCRegionCallbackAlreadyRegistered);
+            }
+
+            EnableNoGCRegionCallbackStatus status = RuntimeImports.RhEnableNoGCRegionCallback(totalSize);
+            if (status != EnableNoGCRegionCallbackStatus.Succeed)
+            {
+                s_NoGCCallback = null;
+                switch (status)
+                {
+                    case EnableNoGCRegionCallbackStatus.NotStarted:
+                        throw new InvalidOperationException(SR.Format(SR.InvalidOperation_UnableToEnableNoGCRegionCallback_NotStarted));
+                    case EnableNoGCRegionCallbackStatus.InsufficientBudget:
+                        throw new InvalidOperationException(SR.Format(SR.InvalidOperation_UnableToEnableNoGCRegionCallback_InsufficientBudget));
+                }
+            }
+            s_NoGCCallback = callback;
+            __Finalizer.s_FinalizerAction = &FinalizerAction;
```

Comment:
We try to maintain separation between Runtime.Base and CoreLib. It would be nice to set the `s_FinalizerAction` via `RuntimeImports` API.